### PR TITLE
Properly format displayName and fields validatation

### LINKED ISSUE #823: Salutation is not suppressed in list view anymore since vue.js
## Expected behavior
If you've decided to sort your contact list based on the first or last name, the salutation should  usually be suppressed in the contact list view, like:
* Sort by first name     -> FIRST LAST
* Sort by last name      -> LAST, FIRT
* Sort by display name   -> SALUTATION FIRST LAST

## Current behavior
Independently which sort order of your contact list you've chosen the display name is always shown.
(I compared the behavior with v2.1.8 in which everything worked as expected)

20190105_nc_contacts_sort_order

### Steps to reproduce
1. Open the Contacts app
2. Switch between the available sort orders and check how the contact names are listed.

#### Server Configuration
OS: Linux 3.16.50
Web server: Apache2 2.4.37
Database: MariaDB 10.2.19
PHP version: 7.2.13
Nextcloud version: 15.0.0
Contacts app version: 3.0.1

#### Client Configuration
Browser: Mozilla Firefox 64.0
Operating system: Windows 10

<bountysource-plugin>

---
Want to back this issue? **[Post a bounty on it!](https://www.bountysource.com/issues/68296159-salutation-is-not-suppressed-in-list-view-anymore-since-vue-js?utm_campaign=plugin&utm_content=tracker%2F46751899&utm_medium=issues&utm_source=github)** We accept bounties via [Bountysource](https://www.bountysource.com/?utm_campaign=plugin&utm_content=tracker%2F46751899&utm_medium=issues&utm_source=github).
</bountysource-plugin>

### LINKED ISSUE #824: placeholder for contact photo shows wrong character in contact list
## Expected behavior
If you've decided to sort your contact list based on the first or last name, the character used
as placeholder for a contact photo, should always be the first character of a displayed name:

* Sort by first name     -> FIRST LAST            -> "F"
* Sort by last name      -> LAST, FIRST           -> "L"
* Sort by display name   -> SALUTATION FIRST LAST -> "S"

## Current behavior
Independently which sort order of your contact list you've chosen the placeholder for a contact photo always contains the first character of the display name. For your information, in the following example the salutation **"Herr"** is used therefore the character **"H"** is displayed although last name sorting has been selected. Instead I would have expect to get an `"A"` displayed because all last names are starting with an **"A"**:

20190105_nc_contact_list_image

### Steps to reproduce
1. Open the Contacts app
2. Switch between the available sort orders and check how the contact names and the placeholders are displayed.

#### Server Configuration
OS: Linux 3.16.50
Web server: Apache2 2.4.37
Database: MariaDB 10.2.19
PHP version: 7.2.13
Nextcloud version: 15.0.0
Contacts app version: 3.0.1

#### Client Configuration
Browser: Mozilla Firefox 64.0
Operating system: Windows 10

<bountysource-plugin>

---
Want to back this issue? **[Post a bounty on it!](https://www.bountysource.com/issues/68297806-placeholder-for-contact-photo-shows-wrong-character-in-contact-list?utm_campaign=plugin&utm_content=tracker%2F46751899&utm_medium=issues&utm_source=github)** We accept bounties via [Bountysource](https://www.bountysource.com/?utm_campaign=plugin&utm_content=tracker%2F46751899&utm_medium=issues&utm_source=github).
</bountysource-plugin>

### LINKED ISSUE #848: Contact name displayed as "New contact" for contacts with only a number of enterprise name
### Describe the bug

When opening the Contacts tab of Nextcloud, all my contacts that are:
- only a phone number, no first or last name
- only a company name and phone number, no first or last name
are displayed as "New contact" in the contact list.

For contacts *with* a company name, when I click on such a contact to view it, the displayed name updates to the company name and the list is updated accordingly.

**To Reproduce**

1. Create a contact with just a phone number or a phone number and a company name
2. (You might need to delete the "New contact" default text in the "Name" field)
3. Refresh the Contacts tab, or go to another tab then come back to it. The new contact should be at the top of the contacts list listed as "New contact".
4. Click on this contact. If you inserted a company name its displayed name should be updated to the company name and the contact will move accordingly in the alphabetical list.

**Expected behavior**

When a contact has a company name and no first/last name, it should be displayed directly as "Company name" and not show up as "New contact" until I click on it.

When a contact only has a phone number, I guess both actions make sense:
- as is: keep it as "new contact" to single it out
- alternative way: some contact managers (iOS in my recollection) use the phone number as display name when no other information is attached to the contact

**Actual behavior**
The contacts were displayed as "New contact".

**Screenshots**
I can take some screenshots if my explanations are not clear enough, I just need to change the display language of my nextcloud instance to make it english and to redact some phone numbers so let me know if it's necessary.

### Server configuration
Operating system: Linux 4.14.66-ovh-vps-grsec-zfs-classid #1 SMP Thu Aug 23 15:15:40 CEST 2018 x86_64

Webserver: Apache (cgi-fcgi)

Database: mysql 5.5.60

PHP version:

7.0.32 Modules loaded: Core, date, libxml, openssl, pcre, sqlite3, zlib, bcmath, bz2, calendar, ctype, curl, dba, dom, hash, fileinfo, filter, ftp, gd, gettext, gmp, SPL, iconv, imagick, session, intl, json, mbstring, mcrypt, memcached, standard, mysqlnd, PDO, pdo_mysql, pdo_pgsql, pdo_sqlite, pgsql, Phar, posix, pspell, redis, Reflection, imap, SimpleXML, soap, sockets, mongodb, exif, sysvmsg, sysvsem, sysvshm, tokenizer, wddx, xml, xmlreader, xmlrpc, xmlwriter, xsl, zip, mysqli, cgi-fcgi, ionCube Loader

Nextcloud version: 14.0.3 - 14.0.3.0

Updated from an older Nextcloud/ownCloud or fresh install: fresh install

Where did you install Nextcloud from: unknown

**Signing status:**
Raw output
==========
Array
(
    [core] => Array
        (
            [INVALID_HASH] => Array
                (
                    [.htaccess] => Array
                        (
                            [expected] => 70183f641f39fc5f43065aeb15b86ac99e86b82310d5467e1c4d935df30acc6155c0905f84f59616885275c19ffacea59e354a87a5c97c94f50d50d4eff0370e
                            [current] => b6802812526e6b12aab5a12d468a59da3b05fde3aacbae5fc465ab8f7f1039a32ee2a2f0e3686ee95aebe0bfef849d2ceb1ce14e2c3fa2fb74f9b4dad39afccd
                        )

                )

        )

)

**List of activated apps:**
Enabled:
 - accessibility: 1.0.1
 - activity: 2.7.0
 - calendar: 1.6.4
 - cloud_federation_api: 0.0.1
 - comments: 1.4.0
 - contacts: 2.1.8
 - dav: 1.6.0
 - federatedfilesharing: 1.4.0
 - federation: 1.4.0
 - files: 1.9.0
 - files_pdfviewer: 1.3.2
 - files_sharing: 1.6.2
 - files_texteditor: 2.6.0
 - files_trashbin: 1.4.1
 - files_versions: 1.7.1
 - files_videoplayer: 1.3.0
 - firstrunwizard: 2.3.0
 - flowupload: 0.0.9
 - gallery: 18.1.0
 - issuetemplate: 0.4.0
 - logreader: 2.0.0
 - lookup_server_connector: 1.2.0
 - nextcloud_announcements: 1.3.0
 - notes: 2.5.1
 - notifications: 2.2.1
 - oauth2: 1.2.1
 - passman: 2.2.1
 - password_policy: 1.4.0
 - provisioning_api: 1.4.0
 - serverinfo: 1.4.0
 - sharebymail: 1.4.0
 - support: 1.0.0
 - survey_client: 1.2.0
 - systemtags: 1.4.0
 - tasks: 0.9.8
 - theming: 1.5.0
 - twofactor_backupcodes: 1.3.1
 - updatenotification: 1.4.1
 - workflowengine: 1.4.0
Disabled:
 - admin_audit
 - encryption
 - files_external
 - user_external
 - user_ldap

**Nextcloud configuration:**
{
    "instanceid": "***REMOVED SENSITIVE VALUE***",
    "passwordsalt": "***REMOVED SENSITIVE VALUE***",
    "secret": "***REMOVED SENSITIVE VALUE***",
    "trusted_domains": [
        "cloud.paulmorin.fr"
    ],
    "datadirectory": "***REMOVED SENSITIVE VALUE***",
    "dbtype": "mysql",
    "version": "14.0.3.0",
    "overwrite.cli.url": "https:\/\/cloud.paulmorin.fr",
    "dbname": "***REMOVED SENSITIVE VALUE***",
    "dbhost": "***REMOVED SENSITIVE VALUE***",
    "dbport": "",
    "dbtableprefix": "oc_",
    "dbuser": "***REMOVED SENSITIVE VALUE***",
    "dbpassword": "***REMOVED SENSITIVE VALUE***",
    "installed": true,
    "memcache.local": "\\OC\\Memcache\\ArrayCache",
    "mail_smtpmode": "smtp",
    "mail_smtpsecure": "ssl",
    "mail_from_address": "***REMOVED SENSITIVE VALUE***",
    "mail_domain": "***REMOVED SENSITIVE VALUE***",
    "mail_smtpauth": 1,
    "mail_smtpauthtype": "LOGIN",
    "mail_smtphost": "***REMOVED SENSITIVE VALUE***",
    "mail_smtpport": "465",
    "mail_smtpname": "***REMOVED SENSITIVE VALUE***",
    "mail_smtppassword": "***REMOVED SENSITIVE VALUE***",
    "debug": true,
    "maintenance": false
}

### Client configuration
Browser: Mozilla/5.0 (Macintosh; Intel Mac OS X 10.13; rv:64.0) Gecko/20100101 Firefox/64.0

Operating system: MacOS 10.13.6

<bountysource-plugin>

---
Want to back this issue? **[Post a bounty on it!](https://www.bountysource.com/issues/68563456-contact-name-displayed-as-new-contact-for-contacts-with-only-a-number-of-enterprise-name?utm_campaign=plugin&utm_content=tracker%2F46751899&utm_medium=issues&utm_source=github)** We accept bounties via [Bountysource](https://www.bountysource.com/?utm_campaign=plugin&utm_content=tracker%2F46751899&utm_medium=issues&utm_source=github).
</bountysource-plugin>

### PR DESCRIPTION
**Pull request by skjnldsv**
Fix #824 
Fix #823
Fix #848

@j-ed @plmrn

Unsigned test build: [contacts.tar.gz](https://github.com/nextcloud/contacts/files/2795686/contacts.tar.gz)

### CODE REVIEW

File: src/models/contact.js [hunk 1]
```diff
@@ -271,25 +273,49 @@ export default class Contact {
...
-				})
-				.join(' ')
+		// BUT if no FN property use the N anyway
+		if (n) {
+			// Stevenson;John;Philip,Paul;Dr.;Jr.,M.D.,A.C.P.
+			// -> John Stevenson
+			return n.slice(0, 2).reverse().join(' ')
+		}
+		// LAST chance, use the org ir that's the only thing we have
+		if (org) {
```
Comment (skjnldsv) [inline src/models/contact.js]: Though this should not happens since FN field is still mandatory.

File: .eslintrc.js [hunk 2]
```diff
@@ -53,7 +53,7 @@ module.exports = {
 		'no-tabs': 0,
 		'vue/html-indent': ['error', 'tab'],
 		// only debug console
-		'no-console': ['error', { allow: ['error', 'warn', 'debug'] }],
+		'no-console': ['error', { allow: ['error', 'warn', 'info', 'debug'] }],
```
Comment (ChristophWurst) [inline .eslintrc.js]: :eyes:

**Comment by plmrn** [comment]
capture d ecran 2019-01-25 a 13 21 09

Tested on Firefox with cleared cache and Safari (which I never use so cache clear as well).

All contacts are not displaying anymore :)

Let me know if you need more info
Thanks a lot

**Comment by plmrn** [comment]
@skjnldsv Yes, and to be fair I think it does take care of the problem, although not exactly the way I intended it a first ;)

I don't know exactly what you mean by "browser logs". Here is the content of Firefox's "console" on the contacts page of the cloud:

> JQMIGRATE: Migrate is installed, version 1.4.0 jquery-migrate.min.js:2:542
> New search handler registered search.js:67:4
> Connected to dav! 
> Object { rootUrl: Getter & Setter, advertisedFeatures: Getter & Setter, currentUserPrincipal: Getter & Setter, principalCollections: Getter & Setter, calendarHomes: Getter & Setter, addressBookHomes: Getter & Setter, parser: Getter & Setter, _isConnected: Getter & Setter, _request: Getter & Setter, … }
> Contacts.vue:238
> TypeError: "r.join is not a function"
> 	get https://cloud.paulmorin.fr/apps/contacts/js/contacts.js:156:2865
> 	sortedContacts https://cloud.paulmorin.fr/apps/contacts/js/contacts.js:112:2432
> 	sortContacts https://cloud.paulmorin.fr/apps/contacts/js/contacts.js:112:2399
> 	We https://cloud.paulmorin.fr/apps/contacts/js/contacts.js:41:33665
> 	commit https://cloud.paulmorin.fr/apps/contacts/js/contacts.js:47:640
> 	commit https://cloud.paulmorin.fr/apps/contacts/js/contacts.js:47:618
> 	_withCommit https://cloud.paulmorin.fr/apps/contacts/js/contacts.js:47:2143
> 	commit https://cloud.paulmorin.fr/apps/contacts/js/contacts.js:47:590
> 	commit https://cloud.paulmorin.fr/apps/contacts/js/contacts.js:41:31431
> 	t https://cloud.paulmorin.fr/apps/contacts/js/contacts.js:91:5248
> 	L https://cloud.paulmorin.fr/apps/contacts/js/contacts.js:276:27677
> 	L https://cloud.paulmorin.fr/apps/contacts/js/contacts.js:276:27545
> 	c https://cloud.paulmorin.fr/apps/contacts/js/contacts.js:276:35404
> addressbooks.js:385:4 

Do you need anything else? Let me know :)

**Comment by skjnldsv** [comment]
@plmrn thanks a lot!
Please try this one (it will still throw an error) but I added an additional debug.
[contacts.tar.gz](https://github.com/nextcloud/contacts/files/2796221/contacts.tar.gz)

**Comment by plmrn** [comment]
Here goes, redacted for privacy purposes (the array of contacts seemed exhaustive before I trimmed it)

> JQMIGRATE: Migrate is installed, version 1.4.0 jquery-migrate.min.js:2:542
> New search handler registered search.js:67:4
> You are running Vue in development mode.
> Make sure to turn on production mode when deploying for production.
> See more tips at https://vuejs.org/guide/deployment.html vue.runtime.esm.js:8096
> Connected to dav! 
> {…}
> ​
> __ob__: Object { value: {…}, dep: {…}, vmCount: 0 }
> ​
> _isConnected: Getter & Setter
> ​
> _request: Getter & Setter
> ​
> addressBookHomes: Getter & Setter
> ​
> advertisedFeatures: Getter & Setter
> ​
> calendarHomes: Getter & Setter
> ​
> currentUserPrincipal: Getter & Setter
> ​
> parser: Getter & Setter
> ​
> principalCollections: Getter & Setter
> ​
> rootUrl: Getter & Setter
> ​
> <prototype>: Object { … }
> Contacts.vue:249
> null contact.js:350
> Array(5) [ "LAST NAME", "FIRST NAME", "", "", "" ]
> contact.js:350
> null contact.js:350
> TypeError: "org.join is not a function"
> 	get https://cloud.paulmorin.fr/apps/contacts/js/contacts.js:43541:18
> 	sortedContacts https://cloud.paulmorin.fr/apps/contacts/js/contacts.js:45500:9
> 	sortContacts https://cloud.paulmorin.fr/apps/contacts/js/contacts.js:45497:8
> 	wrappedMutationHandler https://cloud.paulmorin.fr/apps/contacts/js/contacts.js:41132:5
> 	commitIterator https://cloud.paulmorin.fr/apps/contacts/js/contacts.js:40800:7
> 	commit https://cloud.paulmorin.fr/apps/contacts/js/contacts.js:40799:5
> 	_withCommit https://cloud.paulmorin.fr/apps/contacts/js/contacts.js:40930:3
> 	commit https://cloud.paulmorin.fr/apps/contacts/js/contacts.js:40798:3
> 	boundCommit https://cloud.paulmorin.fr/apps/contacts/js/contacts.js:40743:12
> 	_callee6$/< https://cloud.paulmorin.fr/apps/contacts/js/contacts.js:44921:17
> 	N https://cloud.paulmorin.fr/apps/contacts/js/contacts.js:19545:27674
> 	N https://cloud.paulmorin.fr/apps/contacts/js/contacts.js:19545:27542
> 	u https://cloud.paulmorin.fr/apps/contacts/js/contacts.js:19545:35405
> addressbooks.js:530

**Comment by skjnldsv** [comment]
@plmrn should be fixed in :)
[contacts.tar.gz](https://github.com/nextcloud/contacts/files/2796513/contacts.tar.gz)

**Comment by plmrn** [comment]
Perfect, thanks for the time!

**Pull request merged**